### DISCUSSION TITLE
Creating a state variable with size that can be modified after compilation

### DISCUSSION THREAD

**Discussion by mhpierre** [thelfer/tfel#574] [Q&A] [answered]
Hi Thomas,

I was wondering if it was possible to declare variables and/or material properties of "arbitrary size". What I mean by that is for instance, one could use a `ConstantInteger` to make properties and variables of set size:
```
@ConstantInteger Len = 4;
@MaterialProperty real mat_prop[Len];
@StateVariable real state_var[Len];
```
but as far as I understand the value of `Len` cannot be changed after compilation, or can it?
Is there a way I could set this through a material property or to have , which could make my behaviour highly flexible (being able to handle an arbitrary number `n` of similar mechanisms with associated material properties and state variables)?
I was thinking of using a `std::vector` but wasn't successful at achieving what I wanted.

**Comment by thelfer** [comment, marked as answer]
Hi @LordKthulhu,
No, you can't do that, as it would break many optimisation strategies in MFront.

However, you can do this:

```c++
@DSL Default;
@Behaviour @behaviour@;
@IntegerConstant Len = @nprops@;
@MaterialProperty real mat_prop[Len];
@StateVariable real state_var[Len];

@Integrator{}
```
and compile it this way:

```bash
$ mfront --interface=generic --@behaviour@=example3 --@nprops@=3 test.mfront
$ mfront --interface=generic --@behaviour@=example2 --@nprops@=2 test.mfront
$ mfront --obuild
The following library has been built :
- libBehaviour.so :  example3_AxisymmetricalGeneralisedPlaneStrain example3_Axisymmetrical example3_PlaneStrain example3_GeneralisedPlaneStrain example3_Tridimensional example2_AxisymmetricalGeneralisedPlaneStrain example2_Axisymmetrical example2_PlaneStrain example2_GeneralisedPlaneStrain example2_Tridimensional
```

**Reply by mhpierre** [reply]
I see, it makes sense that it would not be possible. But your alternative is actually pretty compelling and not far at all from what I was imagining, I will definitely make use of it!
Thank you very much.

**Reply by thelfer** [reply]
Note that you can also do this:

```c++
@DSL Default;
@Behaviour example3;
@IntegerConstant Len = 3;
@Import "common.mfront";
```

with `common.mfront` being something like:

```c++
@MaterialProperty real mat_prop[Len];
@StateVariable real state_var[Len];

@Integrator{}
```

Dependending on your use case, writting several top-level MFront files can be more practical than using command-line arguments...

For instance, if you do want your MFront behaviours to be self-contained (see the discussion in this paper https://github.com/thelfer/MFrontGallery/blob/master/docs/papers/joss/paper.md for details), you may want to write:

```c++
@DSL Default;
@Behaviour example3;
@IntegerConstant Len = 3;
@Parameter real mat_props[Len] =  {1, 2, -1};
@Import "common.mfront";
```

(the definition of `mat_props` must then be removed from `common.mfront`)
